### ISSUE TITLE
bazel fetch seems to ignore configs, so can fail due to unused configs.

### ISSUE THREAD

**Issue by viettrungluu** [bazelbuild/bazel#8565]
### Description of the problem / feature request:

bazel fetch seems to ignore configs, so can fail due to unused configs.

For example, gtest has in its (root) BUILD.bazel (https://github.com/google/googletest/blob/8EA412EE54E54CAC6E49EE3B7C209DA1/BUILD.bazel):
```
config_setting(
    name = "has_absl",
    values = {"define": "absl=1"},
)
...
cc_library(
    name = "gtest",
    ....
    deps = select({
        ":has_absl": [
            "@com_google_absl//absl/debugging:failure_signal_handler",
            ...
        ],
        "//conditions:default": [],
    }),
    ...
)
```

If you don't want absl, you don't set the config and you don't put it in your workspace. And everything (bazel build, bazel test) works just fine that way. But bazel fetch fails. Interestingly, bazel sync succeeds.

### Feature requests: what underlying problem are you trying to solve with this feature?

Possibly, this is WAI (and obviously there's an advantage of fetching everything for all configs), but obviously if you don't want absl (for example), you don't want absl. You should still be able to bazel fetch without bringing it in.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

WORKSPACE:
```
load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")

git_repository(
    name = "googletest",
    remote = "https://github.com/google/googletest",
    commit = "8ffb7e5c88b20a297a2e786c480556467496463b",
    shallow_since = "1559187368 -0400",
)
```

BUILD:
```
cc_test(
    name = "hello",
    deps = [
        "@googletest//:gtest",
        "@googletest//:gtest_main",
    ]
)
```

* `bazel fetch //:hello` fails.
* `bazel test //:hello` succeeds.
* `bazel sync` succeeds.

### What operating system are you running Bazel on?

Linux, but pretty sure it doesn't matter.

### What's the output of `bazel info release`?

`release 0.26.0`

###  Have you found anything relevant by searching the web?

Nope.

### Any other information, logs, or outputs that you want to share?

Nope.

**Comment by gregestren** [comment]
I think basing fetch on cquery instead of query could make this work.

**Comment by gregestren** [comment]
But it's true that that wouldn't work as well for people who want to build with multiple configurations.

**Comment by viettrungluu** [comment]
I guess that bazel sync uses cquery (since it works in the above situation)? It's surprising to me that sync works while fetch doesn't, since I'd expect the former to be more complete than the latter.

I tend to favor cquery, since as projects get bigger and more complex, not only do the chances of non-functional configs increase, but the cost of fetching things for all configs also increases (worse, you might not even be able to build some configs locally, thinking about e.g. the Mac/iOS cases).

**Comment by beasleyr-vmw** [comment]
Since Bazel ditched its embedded JDK and java_tools had been factored out, this is much more important, because otherwise Bazel assumes users are able to fetch multiple 100MiB archives for {Win/Lin/macOS}x{unversioned,10,11}x{intel,aarch64}x{openjdk,remotejdk} distributions (see also issue 8268).

We wrap Bazel invocations with a bootstrapping wrapper which implicitly sets HTTP_PROXY and HTTPS_PROXY to prevent users' builds from reaching out to the public internet. (We host the minimum required artifacts in an Artifactory instance and point Bazel to locally available JDKs via custom Java toolchains.) Consequently, `bazel fetch` and `bazel sync` are completely unusable to us. :(

**Comment by TheButlah** [comment]
This is biting me too. At a minimum, there should be a flag that you can give to control whether bazel uses query or cquery. My use case is the exact same one as OP.

Bazel fetch and bazel sync not matching each others semantics in terms of cquery vs query is very strange, but at least the flag can be added without breaking anything.

**Comment by aiuto** [comment]
Everyone is a little bit correct here.

sync should probably continue to do what it advertises, and sync WORKSPACE. 
fetch could be changed to take work off configured targets. That would probably cause no harm to people who are used to fetching without --config=X. 
But fetch could also change to take all branches of selects. I can see a subset of users who would believe that is the correct behavior.

And there are other users who really just want a list of URLs which *would* be fetched, so they can audit that list, download and check into their own repository.  It sounds like vmware does something similar. Certainly Google does. We can't build anything that reaches out.

Further, organizations that do this also need a version of fetch/sync that does the _right thing_(tm) for their users. Their android developers probably only want to fetch the tools needed for android development. But a pure C++ shop might actually want 5 different toolchains only discoverable by config.

Before deciding what path to go down, we should start out with real use cases. I don't want to start from Google's point of view, because our workflows do not involve fetch or sync on a regular basis. @beasleyr-vmw, would you be willing to start a doc describing the workflows you would like your developers to have?

In the meantime, I'm going to reassign this to the ExternalDeps team so they can take a look at it. This is actually their area of responsibility. I think it went to Configurability because it mentioned cquery.

**Comment by beasleyr-vmw** [comment]
> Before deciding what path to go down, we should start out with real use cases. I don't want to start from Google's point of view, because our workflows do not involve fetch or sync on a regular basis. @beasleyr-vmw, would you be willing to start a doc describing the workflows you would like your developers to have?

Sure, I can do that. Would you mind suggesting a venue and format (GitHub gist, Google doc)? Is there a template for this sort of thing? Is this something that other users/orgs would contribute to, or would they fork/provide their own, with you or someone else on the Bazel team collecting the results?

**Comment by aiuto** [comment]
My preference is a world readable google doc so we can do marginal comments. github reviews are not so great for capturing a review. OTOH, this is *your* use case, so there may be less need for review, you can publish it anywhere.

I can't speak for who will collect this. Other people in Bazel are responsible for external dependency tooling. I'm just the person suggesting we find the targets to aim for before we start proposing solutions.

cc @laurentlb

**Comment by beasleyr-vmw** [comment]
@aiuto Sorry for taking so long to get back to this.

https://docs.google.com/document/d/1Z6sVpLrz1GneEDPK_jR2qAd0IicXBALoug43Rp2gXjs/edit?usp=sharing

**Comment by hanshoglund** [comment]
This is also affecting our users: https://github.com/tweag/rules_haskell/issues/1078

**Comment by aiuto** [comment]
Whoops... How did I not see this from May? My bad.
I've add @philwo, whom has recently stepped up to coordinate the many issues related to external deps.

**Comment by sgowroji** [comment]
Hi there! We're doing a clean up of old issues and will be closing this one. Please reopen if you’d like to discuss anything further. We’ll respond as soon as we have the bandwidth/resources to do so.

**Comment by purkhusid** [comment]
I'm the maintainer of rules_dotnet and I'm also encountering this issue there since we heavily use transitions and that requires that the configuration of each target be known. 

Relevant rules_dotnet issue: https://github.com/bazel-contrib/rules_dotnet/issues/476

**Comment by philwo** [comment]
I don't have permission to reopen the issue or unassign / reassign it.

@meteorcloudy Could you have a look please?

Personally, I could never figure out how to use "bazel fetch" correctly, but also have never missed it: My preferred way of ensuring that all required repositories have been downloaded is running "bazel build --nobuild". That works perfectly to fetch only the repositories necessary for my current configuration and targets in my experience.

**Comment by fmeum** [comment]
> over here, that still downloads too much. `bazel build` works, but `bazel build --nobuild` on the same argument fails with some obscure error inside a py_something rule dependency. 

`--nobuild` is meant to perform strictly less work, so this sounds like a strange bug. Can you reproduce this on a standalone project?

**Comment by meteorcloudy** [comment]
I believe the original issue reported should have been fixed by https://github.com/bazelbuild/bazel/commit/c576d5da47948680844168ea6414cc45541525a8.

`bazel build --nobuild` should fetch the exact same external deps as a normal build, if not, please file another issue and provide a minimal reproducible case.

Note that, we also have `bazel vendor` available since Bazel 7, which is also handy to figure out what external dependencies are needed for build specific targets:
https://bazel.build/external/vendor#vendor-target-dependencies
